File: DittoDatto-old/services/mercury-engine/src/mercury_core/models/customer.py
"""Customer model — cross-DB projection from enceladus/users.

Source: schemas/company-blueprint.surql §1.5
Zod origin: packages/shared-types/src/customer.ts

Direction: enceladus → company_X. Never reverse. (Session 9 contract)
This is a CRM view of the customer within a specific company database.
The authoritative user record lives in enceladus/users.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from .common import (
    CustomerChannel,
    CustomerStatus,
    MercuryModel,
    SoftDeleteMixin,
    TimestampMixin,
)


class Customer(MercuryModel, TimestampMixin, SoftDeleteMixin):
    """A customer as seen by a specific company (CRM projection).

    The user_id links back to enceladus/users via string (cross-namespace,
    cannot use record link).
    """

    id: str | None = None

    # Cross-DB reference (string, not record link — different namespace)
    user_id: str | None = None
    store_ids: list[str] = Field(default_factory=list)

    # Profile
    name: str = ""
    first_name: str | None = None
    last_name: str | None = None
    email: str | None = None
    phone: str | None = None
    phone_country_code: str | None = None

    # CRM
    notes: str | None = None
    status: CustomerStatus = CustomerStatus.NEW
    staff_ids: list[str] = Field(default_factory=list)
    channel: CustomerChannel = CustomerChannel.APP

    # Booking refs
    last_booking: str | None = None  # record<booking>

    # Metrics
    total_visits: int = 0
    total_spent: float = 0.0
    first_visit_at: datetime | None = None
    last_visit_at: datetime | None = None

File: DittoDatto-old/packages/shared-types/src/customer.ts
import { z } from "zod";
import { IdSchema, DateTimeSchema } from "./common";

/**
 * Customer Status
 * - new: Just created, no visits yet
 * - active: Has visited within the last 90 days
 * - inactive: No visit in 90+ days
 */
export const CustomerStatusSchema = z.enum(["new", "active", "inactive"]);
export type CustomerStatus = z.infer<typeof CustomerStatusSchema>;

/**
 * The Customer Schema
 * Represents a client of a Business (Company).
 * Has multi-store support via `storeIds` for filtering in the Business Portal.
 * 
 * Inspired by Noona HQ's customer model:
 * - event_count → totalVisits
 * - previous_event → lastBookingId
 * - employee_ids → staffIds
 */
export const CustomerSchema = z.object({
  id: IdSchema,
  companyId: IdSchema,
  storeIds: z.array(IdSchema).default([]), // For multi-store filtering

  // Optional link to a registered Public Marketplace user
  userId: IdSchema.optional(),

  // Profile Details
  name: z.string().min(1).max(255),
  firstName: z.string().optional(),
  lastName: z.string().optional(),
  email: z.string().email().optional().or(z.literal("")),
  phone: z.string().optional(),
  phoneCountryCode: z.string().optional(),
  
  // CRM Data
  notes: z.string().optional(),
  status: CustomerStatusSchema.default("new"),
  staffIds: z.array(IdSchema).default([]), // Staff who have served this customer
  channel: z.enum(["app", "web", "portal", "import"]).default("app"), // Acquisition channel

  // Booking References (Noona pattern: previous_event / next_event)
  lastBookingId: IdSchema.optional(),
  
  // Metrics (incremented by MercuryEngine on booking creation)
  totalVisits: z.number().default(0),
  totalSpent: z.number().default(0), // Cumulative spend in default currency
  firstVisitAt: DateTimeSchema.optional(),
  lastVisitAt: DateTimeSchema.optional(),

  // Metadata
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema,
});

export type Customer = z.infer<typeof CustomerSchema>;

